scripts/bugzilla_attach.py

```diff
@@ -0,0 +1,222 @@
+
+    file_data = ""
+    with open(archive_name, "rb") as data_file:
+        file_data = base64.b64encode(data_file.read())
+
+
+    # Send the data to the target URL (depending on whether using API key or not)
+    if use_api_key:
+        resp = send_data_with_api_key(args.api_key, bug_id, file_data)
+    else:
+        bugzilla_username = input("Enter Bugzilla username: ")
+        bugzilla_password = getpass(prompt="Enter Bugzilla password: ")
+        resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+    resp_json = resp.json()
+
+    # Handle the potential errors
+    while "error" in resp_json:
+        # Using an api key will disable retries, so just output the error message
+        if use_api_key:
+            print(resp_json["message"])
+            break
+        # 300: invalid username or password
+        if resp_json["code"] == 300:
+            print("Incorrect username or password.")
+            bugzilla_username = input("Username (leave blank to exit): ")
+            if bugzilla_username == "":
+                print("Username left blank, exiting")
+                exit()
+            bugzilla_password = getpass(prompt="Password: ")
+            resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+            resp_json = resp.json()
+        # 101: Invalid bug id
+        elif resp_json["code"] == 101:
+            print("Invalid bug id")
+            new_bug_id = input("Enter a new bug id (leave blank to exit): ")
+            if new_bug_id == "":
+                print("ID left blank, exiting")
+                exit()
+            bug_id, valid = try_parse_int(new_bug_id)
+            # Try and see if the new supplied ID is a positive integer
+            while not valid or bug_id <= 0:
+                print("Could not parse bug id as valid, try again")
+                new_bug_id = input("Enter a new bug id (leave blank to exit): ")
+                if new_bug_id == "":
+                    print("ID left blank, exiting")
+                    exit()
+                bug_id, valid = try_parse_int(new_bug_id)
+            resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+            resp_json = resp.json()
+
+
+def try_parse_int(value):
+    """Tries to parse the value as an int"""
+    try:
+        return int(value), True
+    except ValueError:
+        return value, False
+
+def send_data(username, password, bug_id, file_data):
+    """Sends the data to the Bugzilla URL as an attachment"""
+    url = BUGZILLA_URL + 'rest/bug/%s/attachment' % bug_id
+    data = {
+        "login": username,
+        "password": password,
+        "ids": [bug_id],
+        "summary": "Result from must-gather command",
+        "content_type": "application/gzip",
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def send_data_with_api_key(api_key, bug_id, file_data):
```

Comment:
Both functions `send_data` and `send_data_with_api_key` are very similar. Please unify.

Reply:
I'm not so sure unifying them is possible; they may look very similar but the data they are sending has different parameters.

Reply:
You could pass authentication details (api_key or username/password) as dictionary and merge with data dictionary.